modules/tracking/src/trackerStateEstimator.cpp

```diff
@@ -442,4 +442,526 @@ void TrackerStateEstimatorSVM::updateImpl( std::vector<ConfidenceMap>& /*confide
+
+	new_sv->beta = 0.0;
+	new_sv->x = sp;
+	new_sv->x->refCount++;
+	
+	assert(y >= 0);
+	new_sv->y = y;
+	new_sv->g = g;
+
+	supportVectors.push_back(new_sv);
+
+	return new_sv;
+}
+
+void cv::TrackerStateEstimatorStruckSVM::MinGrad(Ptr<SVMSupportPattern> sp, int &y, double &g)
+{
+	y = -1;
+	g = DBL_MAX;
+
+#pragma omp parallel for
+	for (int i = 0; i < (int)sp->rects.size(); i++) {
+		double gradient = -LossFunc(sp->rects[i], sp->rects[sp->y]) - F(sp->x[i], sp->rects[i]);
+		
+		#pragma omp critical
+		{
+			if (gradient < g) {
+				y = i;
+				g = gradient;
+			}
+		}
+	}
+}
+
+void cv::TrackerStateEstimatorStruckSVM::ProcessNew(Ptr<SVMSupportPattern> sp, Ptr<SVMSupportVector> & ypos, Ptr<SVMSupportVector> & yneg)
+{
+	// for positive support vector, loss is zero
+	ypos = AddSupportVector(sp, sp->y, -F(sp->x[sp->y], sp->rects[sp->y]));
+
+	int ymin;
+	double gmin;
+	MinGrad(sp, ymin, gmin);
+
+	yneg = AddSupportVector(sp, ymin, gmin);
+
+	// link between support vectors
+	ypos->relative = Ptr<SVMSupportVector>(yneg);
+	yneg->relative = Ptr<SVMSupportVector>(ypos);
+}
+
+void cv::TrackerStateEstimatorStruckSVM::ProcessOld(Ptr<SVMSupportVector> & ypos, Ptr<SVMSupportVector> & yneg)
+{
+	if (supportPatterns.size() == 0) return;
+
+	// choose ramdomically
+	int ind = rand() % supportPatterns.size();
+	Ptr<SVMSupportPattern> pattern = supportPatterns.at(ind);
+
+	double maxGrad = -DBL_MAX;
+	ypos = Ptr<SVMSupportVector>();
+
+	for (size_t i = 0; i < supportVectors.size(); i++) {		
+		Ptr<SVMSupportVector> sv = supportVectors.at(i);
+
+		if (supportVectors[i]->x != pattern) continue;
+		
+		if ((sv->g > maxGrad) && (sv->beta < (delta(sv->y, pattern->y) * svmC))) {
+			maxGrad = sv->g;
+			ypos = sv;
+		}
+	}
+
+	if (!ypos) return;
+
+	int ymin;
+	double gmin;
+	MinGrad(pattern, ymin, gmin);
+
+	yneg = Ptr<SVMSupportVector>();
+	for (size_t i = 0; i < supportVectors.size(); i++) {
+		Ptr<SVMSupportVector> sv = supportVectors.at(i);
+
+		if (supportVectors.at(i)->x != pattern) continue;
+
+		if (sv->y == ymin) {
+			yneg = sv;
+			break;
+		}
+	}
+
+	if (!yneg)
+		yneg = AddSupportVector(pattern, ymin, gmin);
+}
+
+void cv::TrackerStateEstimatorStruckSVM::Optimize(Ptr<SVMSupportVector> & ypos, Ptr<SVMSupportVector> & yneg)
+{
+	if (supportPatterns.size() == 0) return;
+
+	// choose ramdomically
+	int ind = rand() % supportPatterns.size();
+	Ptr<SVMSupportPattern> pattern = supportPatterns[ind];
+
+	double maxGrad = -DBL_MAX;
+	double minGrad = DBL_MAX;
+
+	ypos = Ptr<SVMSupportVector>();
+	yneg = Ptr<SVMSupportVector>();
+
+	for (int i = 0; i < (int)supportVectors.size(); i++) {
+		const Ptr<SVMSupportVector> sv = supportVectors[i];
+
+		if (sv->x != pattern) continue;
+
+		bool restr = sv->beta < (delta(sv->y, pattern->y) * svmC);
+		
+		if (sv->g > maxGrad && restr) {
+			maxGrad = sv->g;
+			ypos = sv;
+		}
+
+		if (sv->g < minGrad) {
+			minGrad = sv->g;
+			yneg = sv;
+		}
+
+	}
+
+	assert(yneg);
```

Comment:
It's better to use `CV_Assert` macro for this purpose, since it checks the argument regardless of the build mode.